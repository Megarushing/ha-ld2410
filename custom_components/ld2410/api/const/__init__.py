"""Device constants library."""

from __future__ import annotations

from ..enum import StrEnum

DEFAULT_RETRY_COUNT = 3
DEFAULT_RETRY_TIMEOUT = 1
DEFAULT_SCAN_TIMEOUT = 5

CHARACTERISTIC_NOTIFY = "0000fff1-0000-1000-8000-00805f9b34fb"
CHARACTERISTIC_WRITE = "0000fff2-0000-1000-8000-00805f9b34fb"

# ---------- Frame constants (hex strings) ----------
TX_HEADER = "FDFCFBFA"  # Downlink (host→radar) command frame header. Command/ACK use same header/footer.
TX_FOOTER = "04030201"  # Downlink frame footer. ACK frames use this too.  status(2): "0000"=success, "0100"=failure. :contentReference[oaicite:1]{index=1}
RX_HEADER = "F4F3F2F1"  # Uplink (radar→host) data frame header. Types: "01"=engineering, "02"=basic. :contentReference[oaicite:2]{index=2}
RX_FOOTER = "F8F7F6F5"  # Uplink data frame footer.

# NOTE on ACKs: ACK intra-frame data begins with (sent_cmd | 0x0100) then the return payload.
# Example: send "FE00" → ACK contains "FE01" then status(2). :contentReference[oaicite:3]{index=3}


# ---------- Command words (hex strings, little-endian) ----------

# Enable configuration session (must precede other config commands). Returns status + protocol version + buffer size.
CMD_ENABLE_CFG = "FF00"  # value: "0001"
# return: status(2) + proto_ver(2="0001") + buf_size(2="4000"). "0000"=OK, "0100"=fail. :contentReference[oaicite:4]{index=4}

# End configuration session (resume normal working mode).
CMD_END_CFG = "FE00"  # value: (none)
# return: status(2). Send again CMD_ENABLE_CFG to re-enter config mode. :contentReference[oaicite:5]{index=5}

# Set max detection gates (move & still) and "no-one" duration (unoccupied delay).
CMD_SET_MAX_GATES_AND_NOBODY = "6000"
# value (all words little-endian, values are u32 LE):
#   "0000"+<u32 move_gate 2..8>  +  "0100"+<u32 still_gate 2..8>  +  "0200"+<u32 nobody_sec 0..65535>
# return: status(2). Takes effect immediately; retained across power-cycles. Defaults in Table 7. :contentReference[oaicite:6]{index=6}

# Read current configuration (gates, per-gate sensitivities, unoccupied duration).
CMD_READ_PARAMS = "6100"  # value: (none)
# return: status(2) + "AA"(1B) + N_max(1B) + cfg_max_move(1B) + cfg_max_still(1B)
#       + move_sens[0..N](1B each) + still_sens[0..N](1B each) + nobody_duration(2). :contentReference[oaicite:7]{index=7}

# Enable engineering mode (uplink adds per-gate energy arrays, data type "01").
CMD_ENABLE_ENGINEERING = "6200"  # value: (none)
# return: status(2). Setting is volatile (lost on power-off). :contentReference[oaicite:8]{index=8}

# Disable engineering mode (uplink reverts to basic target info, data type "02").
CMD_DISABLE_ENGINEERING = "6300"  # value: (none)
# return: status(2). :contentReference[oaicite:9]{index=9}

# Set sensitivities for a specific gate or all gates.
CMD_SET_SENSITIVITY = "6400"
# value:
#   "0000"+<u32 gate_id 0..8 or 0x0000FFFF for ALL> + "0100"+<u32 move 0..100> + "0200"+<u32 still 0..100>
# return: status(2). Use ALL ("FFFF") to apply uniform values to every gate. :contentReference[oaicite:10]{index=10}

# Read firmware version (type, major, minor/build).
CMD_READ_FW = "A000"  # value: (none)
# return: status(2) + fw_type(2="0001") + major(2) + minor(4). Example decodes to Vx.y.yyyymmddhh. :contentReference[oaicite:11]{index=11}

# Set UART baud rate (persists; takes effect after reboot).
CMD_SET_BAUD = "A100"  # value: BAUD_* index(2)
# return: status(2). Indices: "0001"=9600 … "0008"=460800; factory default "0007"=256000. :contentReference[oaicite:12]{index=12}

# Restore factory defaults (persists; takes effect after reboot).
CMD_FACTORY_RESET = "A200"  # value: (none)
# return: status(2). Table 7 lists default gates/sensitivities/baud. :contentReference[oaicite:13]{index=13}

# Reboot module (module restarts after ACK).
CMD_REBOOT = "A300"  # value: (none)
# return: status(2). :contentReference[oaicite:14]{index=14}

# Bluetooth on/off (requires reboot to apply).
CMD_BT_ONOFF = "A400"  # value: "0001"=ON, "0000"=OFF
# return: status(2). BT is ON by default. :contentReference[oaicite:15]{index=15}

# Get MAC address (over UART).
CMD_GET_MAC = "A500"  # value: "0001"
# return: status(2) + fixed_type(1B="00") + MAC(6B; shown big-endian in example). :contentReference[oaicite:16]{index=16}

# Obtain Bluetooth permission (checks password) — reply is sent via Bluetooth, not UART.
CMD_BT_GET_PERMISSION = (
    "A800"  # value: 6B password ("4869 4C69 6E6B" for "HiLink" split in LE pairs)
)
# return: status(2). Treat "0000"=allowed, "0100"=denied (per generic success/fail). :contentReference[oaicite:17]{index=17}

# Set Bluetooth password (stores new 6B password).
CMD_BT_SET_PWD = "A900"  # value: 6B password (small-end order)
# return: status(2). :contentReference[oaicite:18]{index=18}

# Set distance resolution per gate (0.75 m or 0.2 m; persists; needs reboot).
CMD_SET_RES = "AA00"  # value: RES_* index(2)
# return: status(2). "0000"→0.75 m; "0001"→0.20 m. Example shows index "0001". :contentReference[oaicite:19]{index=19}

# Query distance resolution (returns current index).
CMD_GET_RES = "AB00"  # value: (none)
# return: status(2) + RES_* index(2). Example ACK shows "... 0001 00" → 0.2 m per gate. :contentReference[oaicite:20]{index=20}


# ---------- Parameter words (for 0x0060 “max gates & nobody”) ----------
PAR_MAX_MOVE_GATE = "0000"  # u32 move gate: 2..8
PAR_MAX_STILL_GATE = "0100"  # u32 still gate: 2..8
PAR_NOBODY_DURATION = "0200"  # u32 seconds: 0..65535  (aka "no-one duration"). :contentReference[oaicite:21]{index=21}

# ---------- Parameter words (for 0x0064 “set sensitivity”) ----------
PAR_DISTANCE_GATE = "0000"  # u32 gate: 0..8, or ALL_GATES
PAR_MOVE_SENS = "0100"  # u32 sensitivity: 0..100
PAR_STILL_SENS = "0200"  # u32 sensitivity: 0..100
ALL_GATES = "FFFF"  # special selector meaning "apply to all gates". :contentReference[oaicite:22]{index=22}

# ---------- Baud rate indices (for CMD_SET_BAUD A100) ----------
BAUD_9600 = "0001"
BAUD_19200 = "0002"
BAUD_38400 = "0003"
BAUD_57600 = "0004"
BAUD_115200 = "0005"
BAUD_230400 = "0006"
BAUD_256000 = "0007"  # factory default
BAUD_460800 = "0008"  # per Table 6. :contentReference[oaicite:23]{index=23}

# ---------- Distance resolution indices (for CMD_SET_RES / CMD_GET_RES) ----------
RES_PER_GATE_0_75M = "0000"  # each distance gate = 0.75 m
RES_PER_GATE_0_2M = "0001"  # each distance gate = 0.20 m  (query example returns "0001"). :contentReference[oaicite:24]{index=24}

# ---------- Uplink data types (for RX payload interpretation) ----------
UPLINK_TYPE_ENGINEERING = "01"  # per-gate energies appended to basic target info
UPLINK_TYPE_BASIC = (
    "02"  # basic target info only (default). :contentReference[oaicite:25]{index=25}
)


class Model(StrEnum):
    """Device models."""

    LD2410 = "HLK-LD2410"


__all__ = [
    # exports
    "DEFAULT_RETRY_COUNT",
    "DEFAULT_RETRY_TIMEOUT",
    "DEFAULT_SCAN_TIMEOUT",
    "Model",
    "CHARACTERISTIC_NOTIFY",
    "CHARACTERISTIC_WRITE",
    # frame constants (hex strings)
    "TX_HEADER",
    "TX_FOOTER",
    "RX_HEADER",
    "RX_FOOTER",
    # command words (hex strings, little-endian)
    "CMD_ENABLE_CFG",
    "CMD_END_CFG",
    "CMD_SET_MAX_GATES_AND_NOBODY",
    "CMD_READ_PARAMS",
    "CMD_ENABLE_ENGINEERING",
    "CMD_DISABLE_ENGINEERING",
    "CMD_SET_SENSITIVITY",
    "CMD_READ_FW",
    "CMD_SET_BAUD",
    "CMD_FACTORY_RESET",
    "CMD_REBOOT",
    "CMD_BT_ONOFF",
    "CMD_GET_MAC",
    "CMD_BT_GET_PERMISSION",
    "CMD_BT_SET_PWD",
    "CMD_SET_RES",
    "CMD_GET_RES",
    # parameter words
    "PAR_MAX_MOVE_GATE",
    "PAR_MAX_STILL_GATE",
    "PAR_NOBODY_DURATION",
    "PAR_DISTANCE_GATE",
    "PAR_MOVE_SENS",
    "PAR_STILL_SENS",
    "ALL_GATES",
    # baud rate indices
    "BAUD_9600",
    "BAUD_19200",
    "BAUD_38400",
    "BAUD_57600",
    "BAUD_115200",
    "BAUD_230400",
    "BAUD_256000",
    "BAUD_460800",
    # distance resolution indices
    "RES_PER_GATE_0_75M",
    "RES_PER_GATE_0_2M",
    # uplink data types
    "UPLINK_TYPE_ENGINEERING",
    "UPLINK_TYPE_BASIC",
]


# OLD commands for reference
# CMD_BT_PASS_PRE = b"\xfd\xfc\xfb\xfa\x08\x00\xa8\x00"
# CMD_BT_PASS_DEFAULT = b"HiLink"
# CMD_BT_PASS_POST = b"\x04\x03\x02\x01"
# CMD_ENABLE_CONFIG = b"\xfd\xfc\xfb\xfa\x04\x00\xff\x00\x01\x00\x04\x03\x02\x01"
# CMD_ENABLE_ENGINEERING_MODE = b"\xfd\xfc\xfb\xfa\x02\x00b\x00\x04\x03\x02\x01"
# CMD_DISABLE_CONFIG = b"\xfd\xfc\xfb\xfa\x02\x00\xfe\x00\x04\x03\x02\x01"
#
# MOVING_TARGET = 1
# STATIC_TARGET = 2
#
# frame_start = b"\xf4\xf3\xf2\xf1"
# frame_length = b"(?P<length>..)"
# frame_engineering_mode = b"(?P<engineering>\x01|\x02)"
# frame_head = b"\xaa"
# frame_target_state = b"(?P<target_state>\x00|\x01|\x02|\x03)"
# frame_moving_target_distance = b"(?P<moving_target_distance>..)"
# frame_moving_target_energy = b"(?P<moving_target_energy>.)"
# frame_static_target_distance = b"(?P<static_target_distance>..)"
# frame_static_target_energy = b"(?P<static_target_energy>.)"
# frame_detection_distance = b"(?P<detection_distance>..)"
# frame_engineering_data = b"(?P<engineering_data>.+?)?"
# frame_tail = b"\x55"
# frame_check = b"\x00"
# frame_end = b"\xf8\xf7\xf6\xf5"
#
# frame_maximum_motion_gates = b"(?P<maximum_motion_gates>.)"
# frame_maximum_static_gates = b"(?P<maximum_static_gates>.)"
# frame_motion_energy_gates = b"(?P<motion_energy_gates>.{9})"
# frame_static_energy_gates = b"(?P<static_energy_gates>.{9})"
# frame_additional_information = b"(?P<additional_information>.*)"
#
# frame_regex = (
#         frame_start
#         + frame_length
#         + frame_engineering_mode
#         + frame_head
#         + frame_target_state
#         + frame_moving_target_distance
#         + frame_moving_target_energy
#         + frame_static_target_distance
#         + frame_static_target_energy
#         + frame_detection_distance
#         + frame_engineering_data
#         + frame_tail
#         + frame_check
#         + frame_end
# )
#
# engineering_frame_regex = (
#         frame_maximum_motion_gates
#         + frame_maximum_static_gates
#         + frame_motion_energy_gates
#         + frame_static_energy_gates
#         + frame_additional_information
# )
