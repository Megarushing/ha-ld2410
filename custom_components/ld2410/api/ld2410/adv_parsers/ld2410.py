"""LD2410 parser."""

from __future__ import annotations

def process_ld2410(
    data: bytes | None, mfr_data: bytes | None
) -> dict[str, bool | int]:
    """Process woContact Sensor services data."""
    if data is None and mfr_data is None:
        return {}

    return {
        #TODO: Fill in advertisement data
    }
