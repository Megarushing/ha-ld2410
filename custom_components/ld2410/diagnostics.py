"""Diagnostics support for ld2410 integration."""

from __future__ import annotations

from typing import Any

from homeassistant.components import bluetooth
from homeassistant.components.diagnostics import async_redact_data
from homeassistant.core import HomeAssistant

from .coordinator import LD2410ConfigEntry

TO_REDACT: list[str] = []


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: LD2410ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator = entry.runtime_data

    service_info = bluetooth.async_last_service_info(
        hass, coordinator.ble_device.address, connectable=coordinator.connectable
    )

    return {
        "entry": async_redact_data(entry.as_dict(), TO_REDACT),
        "service_info": service_info,
    }
