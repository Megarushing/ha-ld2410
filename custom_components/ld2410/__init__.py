"""Support for devices."""

import logging

from . import api

from homeassistant.components import bluetooth
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_ADDRESS,
    CONF_MAC,
    CONF_NAME,
    CONF_PASSWORD,
    CONF_SENSOR_TYPE,
    Platform,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers import device_registry as dr

from .const import (
    CONF_RETRY_COUNT,
    CONNECTABLE_MODEL_TYPES,
    DEFAULT_RETRY_COUNT,
    DOMAIN,
    HASS_SENSOR_TYPE_TO_MODEL,
    SupportedModels,
)
from .coordinator import ConfigEntryType, DataCoordinator

PLATFORMS_BY_TYPE = {
    SupportedModels.LD2410.value: [Platform.BINARY_SENSOR, Platform.SENSOR],
}
CLASS_BY_DEVICE = {SupportedModels.LD2410.value: api.LD2410}


_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntryType) -> bool:
    """Set up the device from a config entry."""
    assert entry.unique_id is not None
    if CONF_ADDRESS not in entry.data and CONF_MAC in entry.data:
        # Bleak uses addresses not mac addresses which are actually
        # UUIDs on some platforms (MacOS).
        mac = entry.data[CONF_MAC]
        if "-" not in mac:
            mac = dr.format_mac(mac)
        hass.config_entries.async_update_entry(
            entry,
            data={**entry.data, CONF_ADDRESS: mac},
        )

    if not entry.options:
        hass.config_entries.async_update_entry(
            entry,
            options={CONF_RETRY_COUNT: DEFAULT_RETRY_COUNT},
        )

    sensor_type: str = entry.data[CONF_SENSOR_TYPE]
    model = HASS_SENSOR_TYPE_TO_MODEL.get(sensor_type, api.Model.LD2410)
    # connectable means we can make connections to the device
    connectable = model in CONNECTABLE_MODEL_TYPES
    address: str = entry.data[CONF_ADDRESS]

    await api.close_stale_connections_by_address(address)

    ble_device = bluetooth.async_ble_device_from_address(
        hass, address.upper(), connectable
    )
    if not ble_device:
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="device_not_found_error",
            translation_placeholders={"sensor_type": sensor_type, "address": address},
        )

    cls = CLASS_BY_DEVICE.get(sensor_type, api.Device)
    try:
        device = cls(
            device=ble_device,
            password=entry.data.get(CONF_PASSWORD),
            retry_count=entry.options[CONF_RETRY_COUNT],
        )
    except ValueError as err:
        _LOGGER.error(
            "Device initialization failed because of incorrect configuration parameters: %s",
            err,
        )
        return False

    if entry.data.get(CONF_PASSWORD):
        await device.cmd_send_bluetooth_password()

    coordinator = entry.runtime_data = DataCoordinator(
        hass,
        _LOGGER,
        ble_device,
        device,
        entry.unique_id,
        entry.data.get(CONF_NAME, entry.title),
        connectable,
        model,
    )
    entry.async_on_unload(coordinator.async_start())
    if not await coordinator.async_wait_ready():
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="advertising_state_error",
            translation_placeholders={"address": address},
        )

    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
    await hass.config_entries.async_forward_entry_setups(
        entry, PLATFORMS_BY_TYPE[sensor_type]
    )

    return True


async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Handle options update."""
    await hass.config_entries.async_reload(entry.entry_id)


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    sensor_type = entry.data[CONF_SENSOR_TYPE]
    await entry.runtime_data.device.async_disconnect()
    return await hass.config_entries.async_unload_platforms(
        entry, PLATFORMS_BY_TYPE[sensor_type]
    )
