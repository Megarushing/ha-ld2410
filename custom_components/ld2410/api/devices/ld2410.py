"""Control commands for the device."""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Dict, Sequence

from bleak_retry_connector import BleakClientWithServiceCache

from ..const import (
    CMD_BT_GET_PERMISSION,
    CMD_ENABLE_CFG,
    CMD_END_CFG,
    CMD_ENABLE_ENGINEERING,
    CMD_READ_PARAMS,
    CMD_START_AUTO_THRESH,
    CMD_QUERY_AUTO_THRESH,
    UPLINK_TYPE_BASIC,
    UPLINK_TYPE_ENGINEERING,
)
from .device import Device, OperationError

_LOGGER = logging.getLogger(__name__)


class LD2410(Device):
    """Representation of a device."""

    def __init__(self, *args: Any, **kwargs: Any) -> None:
        """Initialize the device control class."""
        super().__init__(*args, **kwargs)
        self._inverse: bool = kwargs.pop("inverse_mode", False)

    def _disconnected(self, client: BleakClientWithServiceCache) -> None:
        """Handle disconnection and schedule reconnect."""
        super()._disconnected(client)
        if not self._expected_disconnect:
            self.loop.create_task(self._restart_connection())

    async def connect_and_subscribe(self):
        """Begin connection, sends password and enables engineering mode."""
        await self._ensure_connected()
        if self._password_words:
            await self.cmd_send_bluetooth_password()
        await self.cmd_enable_engineering_mode()
        params = await self.cmd_read_params()
        self._update_parsed_data(
            {
                "move_gate_sensitivity": params.get("move_gate_sensitivity"),
                "still_gate_sensitivity": params.get("still_gate_sensitivity"),
                "nobody_duration": params.get("nobody_duration"),
            }
        )

    async def _restart_connection(self) -> None:
        """Reconnect and reauthorize after an unexpected disconnect."""
        try:
            _LOGGER.debug("%s: Reconnecting...", self.name)
            await self.connect_and_subscribe()
        except Exception as ex:  # pragma: no cover - best effort
            _LOGGER.debug("%s: Reconnect failed: %s", self.name, ex)
            await asyncio.sleep(1)
            self.loop.create_task(self._restart_connection())

    async def _execute_timed_disconnect(self) -> None:
        """Execute timed disconnection and schedule reconnect."""
        await super()._execute_timed_disconnect()
        self.loop.create_task(self._restart_connection())

    async def cmd_send_bluetooth_password(
        self, words: Sequence[str] | None = None
    ) -> bool:
        """Send the bluetooth password to the device.

        Returns True if the password is accepted.
        """
        payload_words = words or self._password_words
        if not payload_words:
            raise OperationError("Password required")
        payload = "".join(payload_words)
        key = CMD_BT_GET_PERMISSION + payload
        response = await self._send_command(key)
        if response == b"\x01\x00":
            raise OperationError("Wrong password")
        return response == b"\x00\x00"

    async def cmd_enable_config(self) -> tuple[int, int]:
        """Enable configuration session.

        Returns the protocol version and buffer size.
        """
        response = await self._send_command(CMD_ENABLE_CFG + "0001")
        if not response or len(response) < 6 or response[:2] != b"\x00\x00":
            raise OperationError("Failed to enable configuration")
        proto_ver = int.from_bytes(response[2:4], "little")
        buf_size = int.from_bytes(response[4:6], "little")
        return proto_ver, buf_size

    async def cmd_end_config(self) -> None:
        """End configuration session."""
        response = await self._send_command(CMD_END_CFG)
        if response != b"\x00\x00":
            raise OperationError("Failed to end configuration")

    async def cmd_enable_engineering_mode(self) -> None:
        """Enable engineering mode."""
        await self.cmd_enable_config()
        try:
            response = await self._send_command(CMD_ENABLE_ENGINEERING)
            if response != b"\x00\x00":
                raise OperationError("Failed to enable engineering mode")
        finally:
            await self.cmd_end_config()

    async def cmd_auto_thresholds(self, duration_sec: int) -> None:
        """Start automatic threshold detection for the specified duration."""
        if not 0 <= duration_sec <= 0xFFFF:
            raise ValueError("duration_sec must be 0..65535")
        await self.cmd_enable_config()
        try:
            key = CMD_START_AUTO_THRESH + duration_sec.to_bytes(2, "little").hex()
            response = await self._send_command(key)
            if response != b"\x00\x00":
                raise OperationError("Failed to start automatic threshold detection")
        finally:
            await self.cmd_end_config()

    async def cmd_query_auto_thresholds(self) -> int:
        """Query automatic threshold detection status."""
        response = await self._send_command(CMD_QUERY_AUTO_THRESH)
        if not response or len(response) < 4 or response[:2] != b"\x00\x00":
            raise OperationError("Failed to query automatic threshold status")
        return int.from_bytes(response[2:4], "little")

    async def cmd_read_params(self) -> Dict[str, Any]:
        """Read and parse device configuration parameters."""
        await self.cmd_enable_config()
        try:
            response = await self._send_command(CMD_READ_PARAMS)
            if (
                not response
                or len(response) < 10
                or response[:2] != b"\x00\x00"
                or response[2] != 0xAA
            ):
                raise OperationError("Failed to read parameters")
            payload = response[3:]
            max_gate = payload[0]
            max_move_gate = payload[1]
            max_still_gate = payload[2]
            move_len = max_gate + 1
            expected_len = 3 + move_len * 2 + 2
            if len(payload) < expected_len:
                raise OperationError("Failed to read parameters")
            idx = 3
            move_gate_sensitivity = list(payload[idx : idx + move_len])
            idx += move_len
            still_gate_sensitivity = list(payload[idx : idx + move_len])
            idx += move_len
            nobody_duration = int.from_bytes(payload[idx : idx + 2], "little")
            return {
                "max_gate": max_gate,
                "max_move_gate": max_move_gate,
                "max_still_gate": max_still_gate,
                "move_gate_sensitivity": move_gate_sensitivity,
                "still_gate_sensitivity": still_gate_sensitivity,
                "nobody_duration": nobody_duration,
            }
        finally:
            await self.cmd_end_config()

    def _parse_uplink_frame(self, data: bytes) -> Dict[str, Any] | None:
        """Parse an uplink frame.

        ``data`` must be the payload after removing the frame header and footer.
        Returns ``None`` if the payload is not an uplink frame and raises
        ``ValueError`` if the frame is malformed.
        """
        if len(data) < 2 or data[1] != 0xAA:
            # Not an uplink frame
            return None

        frame_type = data[:1].hex()
        if frame_type == UPLINK_TYPE_ENGINEERING:
            ftype = "engineering"
        elif frame_type == UPLINK_TYPE_BASIC:
            ftype = "basic"
        else:
            raise ValueError(f"unknown frame type {frame_type}")

        if not data.endswith(b"\x55\x00"):
            raise ValueError("missing frame footer")

        content = data[2:-2]
        if len(content) < 9:
            raise ValueError("payload too short for basic data")
        status_raw = content[0]
        move_distance_cm = int.from_bytes(content[1:3], "little")
        move_energy = content[3]
        still_distance_cm = int.from_bytes(content[4:6], "little")
        still_energy = content[6]
        detect_distance_cm = int.from_bytes(content[7:9], "little")
        idx = 9

        moving = status_raw in (0x01, 0x03)
        stationary = status_raw in (0x02, 0x03)
        occupancy = moving or stationary

        result: Dict[str, Any] = {
            "type": ftype,
            "moving": moving,
            "stationary": stationary,
            "occupancy": occupancy,
            "move_distance_cm": move_distance_cm,
            "move_energy": move_energy,
            "still_distance_cm": still_distance_cm,
            "still_energy": still_energy,
            "detect_distance_cm": detect_distance_cm,
        }

        if ftype == "engineering":
            if len(content) < idx + 2:
                raise ValueError("missing gate counts")
            max_move_gate = content[idx]
            max_still_gate = content[idx + 1]
            idx += 2
            move_len = max_move_gate + 1
            still_len = max_still_gate + 1
            if len(content) < idx + move_len + still_len:
                raise ValueError("missing gate energy values")
            move_gate_energy = list(content[idx : idx + move_len])
            idx += move_len
            still_gate_energy = list(content[idx : idx + still_len])
            idx += still_len
            if len(content) < idx + 2:
                raise ValueError("missing photo sensor or OUT pin status")
            photo_sensor = content[idx]
            out_pin = content[idx + 1]
            result.update(
                {
                    "max_move_gate": max_move_gate,
                    "max_still_gate": max_still_gate,
                    "move_gate_energy": move_gate_energy,
                    "still_gate_energy": still_gate_energy,
                    "photo_sensor": photo_sensor,
                    "out_pin": bool(out_pin),
                }
            )

        return result
