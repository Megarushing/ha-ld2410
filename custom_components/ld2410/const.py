"""Constants for the integration."""

from enum import StrEnum

from .api import Model

DOMAIN = "ld2410"
MANUFACTURER = "ld2410"

# Config Attributes

DEFAULT_NAME = "LD2410"


class SupportedModels(StrEnum):
    """Supported models."""

    LD2410 = "ld2410"


CONNECTABLE_MODEL_TYPES: dict[Model, SupportedModels] = {
    Model.LD2410: SupportedModels.LD2410,
}

SUPPORTED_MODEL_TYPES = CONNECTABLE_MODEL_TYPES

HASS_SENSOR_TYPE_TO_MODEL = {str(v): k for k, v in CONNECTABLE_MODEL_TYPES.items()}

# Config Defaults
DEFAULT_RETRY_COUNT = 3

# Config Options
CONF_RETRY_COUNT = "retry_count"
CONF_SAVED_MOVE_SENSITIVITY = "saved_move_gate_sensitivity"
CONF_SAVED_STILL_SENSITIVITY = "saved_still_gate_sensitivity"
