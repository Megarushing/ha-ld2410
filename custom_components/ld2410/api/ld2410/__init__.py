"""Library to handle connection with LD2410."""

from __future__ import annotations

from bleak_retry_connector import (
    close_stale_connections,
    close_stale_connections_by_address,
    get_device,
)

from .adv_parser import LD2410SupportedType, parse_advertisement_data
from .const import (
    DEFAULT_RETRY_COUNT,
    DEFAULT_RETRY_TIMEOUT,
    DEFAULT_SCAN_TIMEOUT,
    LD2410AccountConnectionError,
    LD2410ApiError,
    LD2410AuthenticationError,
    LD2410Model,
)
from .devices.device import (
    LD2410Device,
    LD2410OperationError,
)

from .devices.relay_switch import LD2410RelaySwitch
from .discovery import GetLD2410Devices
from .models import LD2410Advertisement

__all__ = [
    "DEFAULT_RETRY_COUNT",
    "DEFAULT_RETRY_TIMEOUT",
    "DEFAULT_SCAN_TIMEOUT",
    "GetLD2410Devices",
    "LD2410AccountConnectionError",
    "LD2410RelaySwitch",
    "LD2410Advertisement",
    "LD2410ApiError",
    "LD2410AuthenticationError",
    "LD2410Device",
    "LD2410Model",
    "LD2410OperationError",
    "LD2410SupportedType",
    "close_stale_connections",
    "close_stale_connections_by_address",
    "get_device",
    "parse_advertisement_data",
]
