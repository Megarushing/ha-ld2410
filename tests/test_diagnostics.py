"""Tests for the diagnostics data provided by the LD2410 integration."""

from unittest.mock import patch

from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props

from custom_components.ld2410.const import (
    CONF_RETRY_COUNT,
    DEFAULT_RETRY_COUNT,
    DOMAIN,
)
from collections.abc import Callable, Coroutine
from aiohttp.test_utils import TestClient
from homeassistant.config_entries import ConfigEntryState
from homeassistant.const import CONF_ADDRESS, CONF_NAME, CONF_SENSOR_TYPE
from homeassistant.core import HomeAssistant
from typing import Any

from . import LD2410b_SERVICE_INFO

try:
    from tests.common import MockConfigEntry
except ImportError:
    from .mocks import MockConfigEntry

try:
    from tests.components.bluetooth import inject_bluetooth_service_info
except ImportError:
    from .mocks import inject_bluetooth_service_info

try:
    from tests.components.diagnostics import get_diagnostics_for_config_entry
except ImportError:
    from .mocks import get_diagnostics_for_config_entry


async def test_diagnostics(
    hass: HomeAssistant,
    hass_client: Callable[..., Coroutine[Any, Any, TestClient]],
    snapshot: SnapshotAssertion,
) -> None:
    """Test diagnostics for config entry."""

    inject_bluetooth_service_info(hass, LD2410b_SERVICE_INFO)

    with patch(
        "custom_components.ld2410.api.LD2410.update",
        return_value=None,
    ):
        mock_config_entry = MockConfigEntry(
            domain=DOMAIN,
            data={
                CONF_ADDRESS: "AA:BB:CC:DD:EE:FF",
                CONF_NAME: "test-name",
                CONF_SENSOR_TYPE: "ld2410",
            },
            unique_id="aabbccddeeff",
            options={CONF_RETRY_COUNT: DEFAULT_RETRY_COUNT},
            subentries_data=(
                {
                    "title": "Subentry",
                    "data": {},
                    "subentry_type": "test",
                },
            ),
        )
        mock_config_entry.add_to_hass(hass)

        await hass.config_entries.async_setup(mock_config_entry.entry_id)
        await hass.async_block_till_done()
        assert mock_config_entry.state is ConfigEntryState.LOADED

    result = await get_diagnostics_for_config_entry(
        hass, hass_client, mock_config_entry
    )
    assert result == snapshot(
        exclude=props("created_at", "modified_at", "entry_id", "time", "subentry_id")
    )
