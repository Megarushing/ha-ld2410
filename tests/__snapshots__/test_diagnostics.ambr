# serializer version: 1
# name: test_diagnostics
  dict({
    'entry': dict({
      'data': dict({
        'address': '42:6C:99:4F:96:D8',
        'name': 'test-name',
        'sensor_type': 'ld2410',
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'ld2410',
      'minor_version': 1,
      'options': dict({
        'retry_count': 3,
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
        dict({
          'data': dict({
          }),
          'subentry_type': 'test',
          'title': 'Subentry',
          'unique_id': None,
        }),
      ]),
      'title': 'Mock Title',
      'unique_id': '426c994f96d8',
      'version': 1,
    }),
    'service_info': <BluetoothServiceInfoBleak name=HLK-LD2410_96D8 address=42:6C:99:4F:96:D8 rssi=-90 manufacturer_data={256: b'D\x02\x101\x07$\x00Bl\x99O\x96\xd8', 1494: b'\x08\x00JLAISDK'} service_data={} service_uuids=['0000af30-0000-1000-8000-00805f9b34fb'] source=local connectable=True time=0.0 tx_power=-127 raw=None>,
  })
# ---
